src/core/solvers/ipopt/multiple-shooting-nlp.cpp

```diff
@@ -0,0 +1,589 @@
+                                                   ? problem_->get_runningModels()[i]->get_u_lb()(j)
+                                                   : -2e19;
+            x_u[i * (ndx_ + nu_) + ndx_ + j] = problem_->get_runningModels()[i]->get_has_control_limits()
+                                                   ? problem_->get_runningModels()[i]->get_u_ub()(j)
+                                                   : 2e19;
+        }
+    }
+    // Final state bounds
+    for (size_t j = 0; j < ndx_; j++) {
+        x_l[T_ * (ndx_ + nu_) + j] = -2e19;
+        x_u[T_ * (ndx_ + nu_) + j] = 2e19;
+    }
+
+    // Dynamics
+    for (Ipopt::Index i = 0; i < nconst_ - nx_; i++) {
+        g_l[i] = 0;
+        g_u[i] = 0;
+    }
+
+    // Initital conditions
+    for (Ipopt::Index i = 0; i < nx_; i++) {
+        g_l[nconst_ - nx_ + i] = problem_->get_x0()[i];
+        g_u[nconst_ - nx_ + i] = problem_->get_x0()[i];
+    }
+
+    return true;
+}
+// [TNLP_get_bounds_info]
+
+// [TNLP_get_starting_point]
+// returns the initial point for the problem
+bool MultipleShootingNlp::get_starting_point(Ipopt::Index   n,
+                                             bool           init_x,
+                                             Ipopt::Number *x,
+                                             bool           init_z,
+                                             Ipopt::Number *z_L,
+                                             Ipopt::Number *z_U,
+                                             Ipopt::Index   m,
+                                             bool           init_lambda,
+                                             Ipopt::Number *lambda)
+{
+    assert(init_x == true);
+    assert(init_z == false);
+    assert(init_lambda == false);
+
+    // initialize to the given starting point
+    // State variable are always at 0 since they represent increments from the given initial point
+    for (size_t i = 0; i < T_; i++) {
+        for (size_t j = 0; j < ndx_; j++) {
+            x[i * (ndx_ + nu_) + j] = 0;
+        }
+
+        for (size_t j = 0; j < nu_; j++) {
+            x[i * (ndx_ + nu_) + ndx_ + j] = us_[i](j);
+        }
+    }
+
+    for (size_t j = 0; j < ndx_; j++) {
+        x[T_ * (ndx_ + nu_) + j] = 0;
+    }
+
+    return true;
+}
+// [TNLP_get_starting_point]
+
+// [TNLP_eval_f]
+// returns the value of the objective function
+bool MultipleShootingNlp::eval_f(Ipopt::Index n, const Ipopt::Number *x, bool new_x, Ipopt::Number &obj_value)
+{
+    assert(n == nvar_);
+
+    // Running costs
+    for (size_t i = 0; i < T_; i++) {
+        data_.dx      = Eigen::VectorXd::Map(x + i * (ndx_ + nu_), ndx_);
+        data_.control = Eigen::VectorXd::Map(x + i * (ndx_ + nu_) + ndx_, nu_);
+
+        state_->integrate(xs_[i], data_.dx, data_.x);
+
+        problem_->get_runningModels()[i]->calc(problem_->get_runningDatas()[i], data_.x, data_.control);
```

Comment:
It is better to run firs the integration operators over the entire trajectory (and store then in `xs_guess_`).
Then, you call `problem_->calc(xs_guess_, us_guess_)

This approach will run the different computations in parallel, if the user has enabled it.